File: marisaoj/01-binary_search/introduction_to_binary_search/binary_search/main.py
def binary_search(arr: list[int], target: int) -> int:
    l, r = 0, len(arr) - 1

    while l <= r:
        m = l + (r - l) // 2
        if arr[m] == target:
            return m
        if arr[m] < target:
            l = m + 1
        else:
            r = m - 1

    return -1


if __name__ == "__main__":
    n, q = list(map(int, input().split()))
    arr: list[int] = list(map(int, input().split()))
    while q:
        q -= 1
        target = int(input())
        print(binary_search(arr, target) + 1)

File: marisaoj/01-binary_search/introduction_to_binary_search/binary_search/main.cpp
#include <iostream>
#include <vector>

int binary_search(std::vector<int> &arr, int target) {
    int l = 0, r = arr.size() - 1;

    while (l <= r) {
        int m = l  + (r - l) / 2;   
        if (arr[m] == target) 
            return m;

        if (target < arr[m]) 
            r = m - 1;
        
        else 
            l = m + 1;
    }

    return -1; 
}

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(NULL);

    int n, q; 
    std::cin >> n >> q;
    
    std::vector<int> arr(n, 0);
    for (auto &i: arr)
        std::cin >> i;

    while (q--) {
        int target;
        std::cin >> target;
        std::cout << binary_search(arr, target) + 1 << '\n';
    }
}